### ISSUE TITLE
Slightly improve locking approach ChunkSubscriptionManager

### ISSUE THREAD

**Issue by RemcoSpigot** [PureGero/MultiPaper#201]
Currently the ChunkSubscriptionManager locks an entire thread-safe table once accessed or modified. My suggestion would be to first revise this and perhaps only lock when writing to ensure consistency and let the visibility guarantees of the ConcurrentHashMap take care of updating the values. After this a look could be made to improve even further upon this.

Current approach for this PR would be as followed:
1. Naiive implementation of what is currently being done, only with slight improvements
2. Discuss the outcome 
3. If necessary only lock when writing, such as when a lock is acquired or given away
4. Implement any fixes or short comings

**Comment by xymb-endcrystalme** [comment]
Is Master even bottlenecked by this?

**Comment by RemcoSpigot** [comment]
> Is Master bottlenecked by this?

It does seem to have quite the impact when reading chunks. Whenever a single server tries to read a chunk, any other chunks can not be read and possibly have to wait while a chunk owner is being updated, as it tries to load in the current owner.

**Comment by xymb-endcrystalme** [comment]
Oh, it waits for the data from the HDD? In that case yeah, it would be a problem...

On the other hand if that's the case it wouldn't help much as the HDD is a bottleneck anyway.
I assume you're running on an SSD?

**Comment by RemcoSpigot** [comment]
This has nothing to do with any kind of storage, it's in memory. The class keeps track of who is currently holding ownership over the chunk. If you want to be able to read in the data from the Chunk, you have to be owner.

**Comment by RemcoSpigot** [comment]
It looks like GitHub didn't include my commit reference, but here the commit that I made:
https://github.com/RemcoSpigot/MultiPaper/commit/908a6903f758cee41a7d6a209e96c6b1339bd9ab

**Comment by RemcoSpigot** [comment]
Reopening discussion here after closing PR. Approach for the PR was to introduce Reentrant Locking and to open the discussion about reads vs writes. 
Since I don't have too much knowledge on how chunks operate for Spigot, can you elaborate on how this works? How many chunks can be loaded in per server at once? Why 64 chunk locks maximum?

**Comment by PureGero** [comment]
> Since I don't have too much knowledge on how chunks operate for Spigot, can you elaborate on how this works?

Forget Spigot, Paper changes chunk loading completely. MultiPaper then improves the I/O futher.

When a player teleports to a new location, all the chunks around them need to be loaded. First, the server subscribes to these chunks on the master. Then, the server creates an I/O request for each chunk, which is sent to the Master. Once the chunk data is read, the server then either loads or generates the chunk. If the server wants to tick that chunk after it has been loaded, it will ask the master to lock the chunk for them.

> How many chunks can be loaded in per server at once?

Basically, the server will load as many chunks as needed at once unless it gets resource-limited.

> Why 64 chunk locks maximum?

64 was just a random number basically. From testing, it's enough that a server will basically never have to wait to acquire a lock, but also not too many objects that need managing. In theory, 64 is 64x faster than before.

**Comment by PestoNotPasta** [comment]
> 64 was just a random number basically. From testing, it's enough that a server will basically never have to wait to acquire a lock, but also not too many objects that need managing. In theory, 64 is 64x faster than before.

Sorry to necropost but shouldn't Xymb's new chunk format push this limit further?
